accord-core/src/main/java/accord/primitives/Deps.java

```diff
@@ -0,0 +1,978 @@
+                int count = 0;
+                for (int i = 0 ; i < keys.size() ; ++i)
+                {
+                    keyCount += keysToTxnIdCounts[i] > 0 ? 1 : 0;
+                    count += keysToTxnIdCounts[i];
+                }
+                result = new int[count + keyCount];
+            }
+
+            int keyIndex = 0;
+            int offset = keyCount;
+            for (int i = 0 ; i < keys.size() ; ++i)
+            {
+                if (keysToTxnIdCounts[i] > 0)
+                {
+                    int count = keysToTxnIdCounts[i];
+                    int[] src = keysToTxnId[i];
+                    for (int j = 0 ; j < count ; ++j)
+                        result[j + offset] = txnIdMap[src[j]];
+                    Arrays.sort(result, offset, count + offset);
+                    int dups = 0;
+                    for (int j = offset + 1 ; j < offset + count ; ++j)
+                    {
+                        if (result[j] == result[j - 1]) ++dups;
+                        else if (dups > 0) result[j - dups] = result[j];
+                    }
+                    result[keyIndex] = offset += count - dups;
+                    ++keyIndex;
+                }
+            }
+            if (offset < result.length)
+                result = Arrays.copyOf(result, offset);
+
+            Keys keys = this.keys;
+            if (keyCount < keys.size())
+            {
+                keyIndex = 0;
+                Key[] newKeys = new Key[keyCount];
+                for (int i = 0 ; i < keys.size() ; ++i)
+                {
+                    if (keysToTxnIdCounts[i] > 0)
+                        newKeys[keyIndex++] = keys.get(i);
+                }
+                keys = new Keys(newKeys);
+            }
+
+            return new Deps(keys, txnIds, result);
+        }
+    }
+
+    public static Builder builder(Keys keys)
+    {
+        return new Builder(keys);
+    }
+
+    static class MergeStream
+    {
+        final Deps source;
+        // TODO: could share backing array for all of these if we want, with an additional offset
+        final int[] input;
+        final int keyCount;
+        int[] remap; // TODO: use cached backing array
+        int[] keys; // TODO: use cached backing array
+        int keyIndex;
+        int index;
+        int endIndex;
+
+        MergeStream(Deps source)
+        {
+            this.source = source;
+            this.input = source.keyToTxnId;
+            this.keyCount = source.keys.size();
+        }
+
+        private void init(Keys keys, TxnId[] txnIds)
+        {
+            this.remap = remapper(source.txnIds, txnIds, true);
+            this.keys = source.keys.remapper(keys, true);
+            while (input[keyIndex] == keyCount)
+                ++keyIndex;
+            this.index = keyCount;
+            this.endIndex = input[keyIndex];
+        }
+    }
+
+    public static <T> Deps merge(Keys keys, List<T> merge, Function<T, Deps> getter)
```

Comment:
Switching `Keys` to be 

```
Keys keys = list.stream().map(d -> d.test.keys()).reduce(Keys.EMPTY, (l, r) -> Keys.union(l, r));
```

gets `merge` to no longer fail, so was able to get my hacked up benchmarks (pushed to review/minimal-deps but should remove after review is done)

```
Benchmark[name=merge(size=6, keys=279)] avg 131.05ms, allocated 5357 MiB
Benchmark[name=builder forEach(size=6, keys=279)] avg 398.75ms, allocated 9418 MiB
Benchmark[name=merge(size=14, keys=300)] avg 555.55ms, allocated 16332 MiB
Benchmark[name=builder forEach(size=14, keys=300)] avg 898.02ms, allocated 18628 MiB
Benchmark[name=merge(size=11, keys=320)] avg 343.21ms, allocated 11180 MiB
Benchmark[name=builder forEach(size=11, keys=320)] avg 736.06ms, allocated 15173 MiB
Benchmark[name=merge(size=15, keys=342)] avg 508.55ms, allocated 16129 MiB
Benchmark[name=builder forEach(size=15, keys=342)] avg 919.88ms, allocated 19336 MiB
Benchmark[name=merge(size=16, keys=355)] avg 689.76ms, allocated 22904 MiB
Benchmark[name=builder forEach(size=16, keys=355)] avg 1,162.81ms, allocated 22445 MiB
```

I do see that if you are in a loop merging `Deps` non stop that `merge` is faster but the amount of allocated memory isn't that much smaller than the `Builder` way

Reply:
merge currently has a handful of trivially cacheable large allocations, so it will be much faster and lower allocation when those simple TODO are completed. It was more important to me to do the difficult work first, but you’re welcome to tackle the TODOs as part of review to make it faster if you like! (Probably easier once we have imported c* utils)

Reply:
> welcome to tackle the TODOs as part of review to make it faster if you like

Right now I am focused on correctness rather than performance, so was hoping that there wasn't a clear win so I could argue to remove in favor of simpler code.  I made an op in builder to allocate less, so now `builder` allocates less than `merge`, but `merge` is still ~30% faster.

Just because I already had everything in place, here is a flame graph of both code paths:

merge
<img width="1709" alt="Screen Shot 2022-08-09 at 8 50 57 AM" src="https://user-images.githubusercontent.com/1852601/183699251-8004e511-e518-43f7-a8eb-0afbf48d5ff7.png">

builder
<img width="1722" alt="Screen Shot 2022-08-09 at 8 51 21 AM" src="https://user-images.githubusercontent.com/1852601/183699355-df064618-3781-41d1-8925-694a2ee0c13c.png">

Reply:
Merge will allocate close to nothing when it uses caching, there is zero change builder will be able to allocate less.

Reply:
yep, my tests bias towards large number of keys (200+) but when smaller there really isn't a difference between the two.

Just pushed a property test to validate merge and builder "do the right thing", and this shows that merge currently doesn't work with partial keys; trying to figure out why now.

Reply:
Wdym partial keys?

The number of deps we are merging, the number of txns per key, and the amount of overlap between the deps  and distribution of that overlap all matter a great deal. Again, a performance comparison should also only really be made with the simple caching implemented (for a test this could be as simple as a thread local of the maximum allocated size of each buffer)

Reply:
Oh, gosh, there's a huge amount of performance left on the table. We're not efficiently merging `TxnId` (the TODO depends on a cassandra-utils jar or copy/pasting `MergeIterator`), let alone optimisations to `remapper`.

So far the only important work done is optimally merging the `int[]`, none of the other setup (which is why it is only 1/3rd of the time at present)

I assume your test generates entirely random keys and `TxnId`? Do you have a working branch?

Reply:
> Wdym partial keys?

So, it looks like you mean if you were to provide a `Keys` object that does not contain all of the keys of the lists you are merging. In this case, the method is not expected to succeed. We can document that, or if you prefer I can modify it to remove data for the Keys that are missing.

Reply:
> In this case, the method is not expected to succeed

the usage of `merge` looks like the following at all call paths

```
Deps.merge(txn.keys, recoverOks, ok -> ok.earlierAcceptedNoWitness)
```

so always taking the `oks.deps` and merging.  Based off what I think you are saying, the `txn.keys` has a hard requirement that it contains every key found in `oks.deps.keys`?  If so a bug in the `oks` (such as we forget to strip out non intersecting keys) logic would then cause the coordinator to fail during `merge` rather than remove any unreferenced keys?  So the desired behavior in this case is not to "merge" the Deps but instead to reject the input?

Reply:
> a performance comparison should also only really be made with the simple caching implemented

Sorry for causing any confusion for motivations behind the perf tests I hacked together, this was to help me try to understand the motivations behind the code.  The code is dense and not documented, so "what should it do" is something I infer based off the name, so I saw another way to handle with `Builder` so I wanted to ask "what is the gain we see"; this work is meant to be thrown away and was to help me justify the time to review rather than replace.

Reply:
> If so a bug in the oks (such as we forget to strip out non intersecting keys) logic would then cause the coordinator to fail during merge rather than remove any unreferenced keys?

If there's a bug in the `ok` response all bets are off for correctness, so yeah we probably want to fail if it somehow includes an unexpected key (and in fact if you like we can optionally enforce this in `remapper`)

Reply:
cool, will correct the test and try to have this detected and produce a cleaner error

Reply:
pushed a change to 574bcd14cfce7073787ea430e0db79b01bad76b7 to detect this in remapper when targetSuperset=true, I figured I will keep the old behavior when its false; all call paths are true today

Reply:
> my tests bias towards large number of keys (200+) but when smaller there really isn't a difference between the two.

Interesting. I just checked out your PR and made some tweaks to improve `remapper`, use `MergeIterator` and cache `int[]`, and while the performance is relatively stable for large numbers of keys (I haven't compared for small numbers of keys), I see a performance delta that is much larger between the two approaches for smaller numbers of keys.

```
13:22:43.148 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=2, keys=7, size=904)] avg 7.65ms, 0.21ms GC, allocated 594 MiB
13:22:45.846 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=2, keys=7, size=904)] avg 63.79ms, 0.22ms GC, allocated 3289 MiB
13:22:45.872 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=1, keys=14, size=149)] avg 0.06ms, 0.17ms GC, allocated 4 MiB
13:22:46.145 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=1, keys=14, size=149)] avg 6.13ms, 0.21ms GC, allocated 340 MiB
13:22:57.425 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=11, keys=21, size=3863)] avg 249.98ms, 0.29ms GC, allocated 3297 MiB
13:23:12.559 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=11, keys=21, size=3863)] avg 359.93ms, 0.23ms GC, allocated 11533 MiB
13:23:15.202 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=3, keys=13, size=1512)] avg 56.32ms, 0.23ms GC, allocated 1400 MiB
13:23:21.237 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=3, keys=13, size=1512)] avg 143.27ms, 0.22ms GC, allocated 5531 MiB
```

Reply:
For comparison, I included running `with` in a loop, and to my surprise this is often faster. It allocates a lot more garbage, but this could readily be optimised by separating `with` into a function that operates on provided in/out arrays, and re-using pairs of buffers.

Given `merge` has lower complexity than `with` it suggests we're leaving some constant factors on the table. Since it's faster even with only 4 streams to merge, it's not as simple as cache associativity, however optimally we might anyway prefer to buffer the input streams in a cache-sympathetic manner. It's possible that a sequence of linear merges has better pipelining / predictability.

It's worth noting that in the real world, we expect to mostly merge either disjoint keys or keys with `txnId` lists that are mostly the same. So, we should run some tests specifically with this in mind to verify approaches. We don't anticipate lots of keys, but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

Either way, perhaps it would be fine to simplify `merge` into a variant of recursive `with` that produces less garbage, and later perhaps introduce a variant that balances algorithmic complexity (which is very poor with this approach) and constant factors.


```
13:47:29.482 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=4, keys=21, size=1646)] avg 168.35ms, 0.22ms GC, allocated 6901 MiB
13:47:31.980 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=4, keys=21, size=1646)] avg 59.03ms, 0.24ms GC, allocated 1763 MiB
13:47:33.485 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=4, keys=21, size=1646)] avg 35.4ms, 0.22ms GC, allocated 2308 MiB
13:48:15.296 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=15, keys=27, size=8449)] avg 951.11ms, 0.23ms GC, allocated 30617 MiB
13:48:48.707 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=15, keys=27, size=8449)] avg 795.14ms, 0.22ms GC, allocated 13242 MiB
13:49:16.317 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=15, keys=27, size=8449)] avg 656.92ms, 0.23ms GC, allocated 35741 MiB
13:49:44.247 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=9, keys=22, size=5812)] avg 638.99ms, 0.24ms GC, allocated 22966 MiB
13:50:01.518 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=9, keys=22, size=5812)] avg 410.77ms, 0.23ms GC, allocated 8037 MiB
13:50:14.952 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=9, keys=22, size=5812)] avg 319.43ms, 0.23ms GC, allocated 16794 MiB
```

Reply:
> It's worth noting that in the real world, we expect to mostly merge either disjoint keys or keys with txnId lists that are mostly the same. So, we should run some tests specifically with this in mind to verify approaches. We don't anticipate lots of keys, but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

This makes a lot of sense, if you happen to touch the same key vs you are the same txn just with slightly different inputs so all keys conflict.

> but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

yeah, also makes sense, use smaller number of keys with larger `Deps` list.

Reply:
> perform the efficient with variation ... the simplicity is probably well justified.

WFM, ill focus on `with` and `without` and ignore `merge` for now